src/libre/re_char_class.h

```diff
@@ -0,0 +1,182 @@
+#ifndef RE_CHAR_CLASS_H
+#define RE_CHAR_CLASS_H
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <assert.h>
+#include <string.h>
+
+#include <fsm/fsm.h>
+#include <fsm/bool.h>
+#include <fsm/pred.h>
+
+#include <re/re.h>
+
+/* This is a duplicate of struct lx_pos, but since we're linking to
+ * code with several distinct lexers, there isn't a clear lexer.h
+ * to include here. The parser sees both definitions, and will
+ * build a `struct ast_pos` in the appropriate places. */
+struct ast_pos {
+	unsigned byte;
+	unsigned line;
+	unsigned col;
+};
+
+
+/* typedef for character class constructors whose prototypes
+ * appear in class.h -- this typedef should be moved there
+ * later. */
+typedef struct fsm *
+char_class_constructor_fun(const struct fsm_options *opt);
+
+/* For a particular regex dialect, this function will look up
+ * whether a name refers to a known character class, and if so,
+ * return the constructor function for it.
+ * Returns whether lookup suceeded. */
+typedef int
+re_dialect_char_class_fun(const char *name, char_class_constructor_fun **res);
+
+enum re_char_class_flags {
+	RE_CHAR_CLASS_FLAG_NONE = 0x00,
+	/* the class should be negated, e.g. [^aeiou] */
+	RE_CHAR_CLASS_FLAG_INVERTED = 0x01,
+	/* includes the `-` character, which isn't part of a range */
+	RE_CHAR_CLASS_FLAG_MINUS = 0x02
+};
+
+enum re_char_class_ast_type {
+	RE_CHAR_CLASS_AST_CONCAT,
+	RE_CHAR_CLASS_AST_LITERAL,
+	RE_CHAR_CLASS_AST_RANGE,
+	RE_CHAR_CLASS_AST_NAMED,
+	RE_CHAR_CLASS_AST_CHAR_TYPE,
+	RE_CHAR_CLASS_AST_FLAGS,
+	RE_CHAR_CLASS_AST_SUBTRACT
+};
+
+/* These are used for named character classes, such as [:alnum:]. */
+enum ast_char_class_id {
+	AST_CHAR_CLASS_ALNUM,
+	AST_CHAR_CLASS_ALPHA,
+	AST_CHAR_CLASS_ANY,
+	AST_CHAR_CLASS_ASCII,
+	AST_CHAR_CLASS_BLANK,
+	AST_CHAR_CLASS_CNTRL,
+	AST_CHAR_CLASS_DIGIT,
+	AST_CHAR_CLASS_GRAPH,
+	AST_CHAR_CLASS_LOWER,
+	AST_CHAR_CLASS_PRINT,
+	AST_CHAR_CLASS_PUNCT,
+	AST_CHAR_CLASS_SPACE,
+	AST_CHAR_CLASS_SPCHR,
+	AST_CHAR_CLASS_UPPER,
+	AST_CHAR_CLASS_WORD,
+	AST_CHAR_CLASS_XDIGIT
+};
+
+enum ast_range_endpoint_type {
+	AST_RANGE_ENDPOINT_LITERAL,
+	AST_RANGE_ENDPOINT_CHAR_TYPE,
+	AST_RANGE_ENDPOINT_CHAR_CLASS
+};
+struct ast_range_endpoint {
+	enum ast_range_endpoint_type t;
+	union {
+		struct {
+			unsigned char c;
+		} literal;
+		struct {
+			char_class_constructor_fun *ctor;
+		} char_class;
+	} u;
+};
+
+struct re_char_class_ast {
+	enum re_char_class_ast_type t;
+	union {
+		struct {
+			struct re_char_class_ast *l;
+			struct re_char_class_ast *r;
+		} concat;
+		struct {
+			unsigned char c;
+		} literal;
+		struct {
+			struct ast_range_endpoint from;
+			struct ast_pos start;
+			struct ast_range_endpoint to;
+			struct ast_pos end;
+		} range;
+		struct {
+			char_class_constructor_fun *ctor;
+		} named;
+		struct {
+			enum re_char_class_flags f;
+		} flags;
+		struct {
+			struct re_char_class_ast *ast;
+			struct re_char_class_ast *mask;
+		} subtract;
+	} u;
+};
```

Comment:
Probably also after this settles, I would like to consider making `struct re_char_class_ast` less tree-shaped - I think we can get rid of the `.concat` node and have classes as an array of `enum re_char_class_ast_type` items.